UCP/IB/CUDA: Direct NIC support

tvegas1 opened the pull request:
## What?
Add direct NIC support.

yosefe commented on src/uct/api/uct.h:
```diff
@@ -1608,7 +1608,15 @@ typedef enum uct_md_mem_attr_field {
+     * region, and can be used to register the region with another memory
+     * domain that need an explicit PCIe BAR window.
+     */
+    UCT_MD_MEM_ATTR_FIELD_DMABUF_FD_PCIE = UCS_BIT(6)
```
can we do this without adding DMABUF_FD_PCIE to the API, and reusing the existing dmabuf_fd to expose the direct NIC mode?

tvegas1 replied:
agreed to fix is subsequent pr

tvegas1 commented on src/uct/cuda/cuda_copy/cuda_copy_md.c:
```diff
@@ -878,17 +928,33 @@ static int uct_cuda_copy_md_get_dmabuf_fd(uintptr_t address, size_t length)
 #endif
 
+    if (explicit_pci_export) {
+        if (!uct_cuda_copy_md_has_c2c() || !pcie_export_flags) {
```
fixed

tvegas1 commented on src/uct/ib/mlx5/dv/ib_mlx5dv_md.c:
```diff
@@ -2171,6 +2247,46 @@ static void uct_ib_mlx5dv_check_dm_ksm_reg(uct_ib_mlx5_md_t *md)
+        md->direct_nic_sys_dev = UCS_SYS_DEVICE_ID_UNKNOWN;
+    }
+
+    ucs_debug("%s: Direct NIC %s supported ret=%d sys_path='%s%s' "
```
fixed

yosefe commented on src/ucp/rma/put_offload.c:
```diff
@@ -134,7 +135,13 @@ ucp_proto_put_offload_bcopy_send_func(ucp_request_t *req,
-    return ucp_proto_bcopy_send_func_status(packed_size);
+    status = ucp_proto_bcopy_send_func_status(packed_size);
+    if ((status == UCS_INPROGRESS) || (status == UCS_OK)) {
+        ucp_ep_flush_mem_schedule(req, lpriv->super.lane,
```
1. do we also need it for put_offload_short, and the fast-path put short flow?
2. can we reduce overhead of this operation?

tvegas1 replied:
1. i think so, 2. maybe, will address in flush-related PR

yosefe commented on src/ucs/sys/topo/base/topo.h:
```diff
@@ -45,13 +45,20 @@ typedef struct ucs_sys_bus_id {
 
+enum {
+    /* Memory's system device needs flush */
+    UCS_SYS_DISTANCE_NEEDS_FLUSH = UCS_BIT(0)
```
why is it in topo.h, i don't see any function in topo.c that returns this flag?

yosefe commented on src/uct/ib/configure.m4:
```diff
@@ -282,11 +282,19 @@ AS_IF([test "x$with_ib" = "xyes"],
+                           mlx5dv_reg_dmabuf_mr], [], [have_direct_nic=no],
+                          [[#include <infiniband/mlx5dv.h>]])
+           AS_IF([test "x$have_direct_nic" != xno],
+               [AC_DEFINE([HAVE_DIRECT_NIC], 1, [Have Direct NIC support])])])
```
do we really need this HAVE_DIRECT_NIC macro? maybe can use the ones generated automatically from AC_CHECK_DECLS line above

yosefe replied:
we can keep both since they are enabling independent flows in the code. query controls setting md->direct_nic_sys_dev and reg_mr is when registration happens. each macro can be used independently.

yosefe commented on src/uct/ib/mlx5/dv/ib_mlx5dv_md.c:
```diff
@@ -750,12 +814,24 @@ uct_ib_mlx5_devx_reg_mr(uct_ib_mlx5_md_t *md, uct_ib_mlx5_devx_mem_t *memh,
+    status = uct_ib_mlx5_direct_nic_reg_mr(md, address, length,
+                                           params, access_flags,
+                                           &memh->mrs[mr_type].super.ib);
+    if (status == UCS_ERR_INVALID_PARAM) {
+        return status;
+    }
```
seems unfixed

tvegas1 replied:
fixed

yosefe commented on src/uct/ib/mlx5/dv/ib_mlx5dv_md.c:
```diff
@@ -714,6 +714,64 @@ uct_ib_mlx5_devx_symmetric_rkey(const uct_ib_mlx5_md_t *md, unsigned flags)
+                              struct ibv_mr **mr_p)
+{
+#ifdef HAVE_DIRECT_NIC
+    int dmabuf_fd_pcie;
```
rename to dmabuf_fd

yosefe commented on src/uct/ib/mlx5/dv/ib_mlx5dv_md.c:
```diff
@@ -2171,6 +2241,57 @@ static void uct_ib_mlx5dv_check_dm_ksm_reg(uct_ib_mlx5_md_t *md)
+                                                     md->direct_nic_sys_dev);
+        ucs_assertv_always(status == UCS_OK, "set sys_dev_aux failed: %s",
+                           ucs_status_string(status));
+        md->super.dev.flags |= UCT_IB_DEVICE_FLAG_DIRECT_NIC;
```
maybe no need for this flag, can check md->direct_nic_sys_dev!=UNKNOWN

tvegas1 replied:
removed

yosefe commented on src/uct/ib/mlx5/dv/ib_mlx5dv_md.c:
```diff
@@ -2171,6 +2241,57 @@ static void uct_ib_mlx5dv_check_dm_ksm_reg(uct_ib_mlx5_md_t *md)
+    ret = mlx5dv_get_data_direct_sysfs_path(ctx, sys_path, sizeof(sys_path));
+    if (ret == 0) {
+        /* Create a DMA specific device from topology perspective */
+        snprintf(dev_name, sizeof(dev_name), "%s_dma",
```
why need _dma suffix? it can change UCX_NET_DEVICES behavior

tvegas1 replied:
tbh now I think we might even be able to not allocate an aux_sysdev, and just add auxiliary path to existing device. aux_sys_dev is only used in topo code, wdyt?

yosefe replied:
IMO better keep the 1:1 mapping between sys_dev and BDF/sysfs path

tvegas1 replied:
fixed on #10760

yosefe commented on src/uct/ib/mlx5/dv/ib_mlx5dv_md.c:
```diff
@@ -2171,6 +2241,57 @@ static void uct_ib_mlx5dv_check_dm_ksm_reg(uct_ib_mlx5_md_t *md)
+        md->super.dev.flags |= UCT_IB_DEVICE_FLAG_DIRECT_NIC;
+    } else {
+        sys_path[0]            = '\0';
+        md->direct_nic_sys_dev = UCS_SYS_DEVICE_ID_UNKNOWN;
```
set it outside of HAVE_DIRECT_NIC

tvegas1 replied:
added outside

yosefe commented on src/uct/ib/mlx5/dv/ib_mlx5dv_md.c:
```diff
@@ -2171,6 +2241,57 @@ static void uct_ib_mlx5dv_check_dm_ksm_reg(uct_ib_mlx5_md_t *md)
+                                                  md->direct_nic_sys_dev);
+    }
+
+    ucs_debug("%s: Direct NIC %s supported ret=%d sys_path='%s%s' "
```
smth like 
ret = mlx5dv_get_data_direct_sysfs_path ...
if (ret != 0) {
ucs_debug("%s: mlx5dv_get_data_direct_sysfs_path(%s failed: %m", ib_dev_name, sys_path...)
return
}

ucs_debug("%s: direct nic is supported, sys_dev %s", ib_dev_name, sys_dev_name")

yosefe commented on src/uct/ib/mlx5/dv/ib_mlx5dv_md.c:
```diff
@@ -2370,6 +2491,8 @@ ucs_status_t uct_ib_mlx5_devx_md_open_common(const char *name, size_t size,
 
     uct_ib_mlx5_devx_check_odp(md, md_config, cap);
 
+    uct_ib_mlx5dv_check_direct_nic(ctx, dev, md, md_config);
```
need it also for uct_ib_mlx5dv_md_open (mlx5dv non-devx)

tvegas1 commented on src/ucp/core/ucp_mm.c:
```diff
@@ -598,6 +626,11 @@ ucp_memh_register_internal(ucp_context_h context, ucp_mem_h memh,
             reg_params.field_mask |= UCT_MD_MEM_REG_FIELD_DMABUF_FD |
                                      UCT_MD_MEM_REG_FIELD_DMABUF_OFFSET;
+            sys_dev_map            = context->tl_mds[md_index].sys_dev_map;
+            if (!ucp_memh_sys_dev_reachable(md_index, mem_attr.sys_dev,
```
fixed

tvegas1 commented on src/ucp/core/ucp_ep.h:
```diff
@@ -492,14 +492,35 @@ struct ucp_ep_config {
+         * Memory areas that need specific flushing, tracked by device as there
+         * could be many memory devices behind a given lane.
+         */
+        ucp_mem_area_t entry[UCS_SYS_DEVICE_ID_MAX];
```
agreed, needs hash eventually

tvegas1 commented on src/ucp/core/ucp_context.c:
```diff
@@ -1271,6 +1276,7 @@ ucp_add_tl_resources(ucp_context_h context, ucp_md_index_t md_index,
 
     /* copy only the resources enabled by user configuration */
     context->tl_rscs = tmp;
+    md->sys_dev_map  = 0;
```
makes sense fixed

tvegas1 commented on src/ucp/core/ucp_rkey.c:
```diff
@@ -611,11 +653,27 @@ int ucp_memh_buffer_is_dummy(const void *exported_memh_buffer)
 {
-    ucp_context_h context = memh->context;
+    ucp_context_h context         = memh->context;
+    ucp_sys_dev_map_t sys_dev_map = ucp_memh_md_map_sys_dev_map(context,
```
will address in flush-related PR

tvegas1 commented on src/ucp/core/ucp_rkey.c:
```diff
@@ -104,13 +116,33 @@ void ucp_rkey_packed_copy(ucp_context_h context, ucp_md_map_t md_map,
+    ucs_sys_dev_distance_t tmp;
+    double latency_nsec;
+
+    if (distance == NULL) {
```
discussion for flush-related PR

tvegas1 commented on src/ucp/core/ucp_rkey.c:
```diff
@@ -70,7 +71,18 @@ size_t ucp_rkey_packed_size(ucp_context_h context, ucp_md_map_t md_map,
+            size += sizeof(uint8_t);
+
+            /* Distance of each device */
+            size += ucs_popcount(sys_dev_map) *
```
to be addressed in flush-related PR

tvegas1 commented:
@yosefe, addressed all current comments, and fixed merge conflict, we can continue review of datapath part on https://github.com/openucx/ucx/pull/10760

tvegas1 commented on src/ucp/core/ucp_request.h:
```diff
@@ -335,6 +346,8 @@ struct ucp_request {
                     uint8_t            num_lanes; /* How many lanes are being flushed */
                     ucp_lane_map_t     started_lanes; /* Which lanes need were flushed */
+
+                    ucp_mem_flush_t    mem;       /* Memory specific flushes */
```
double checked, it adds 8 bytes to `ucp_request_t` if we reorder perfectly this overall flush section/structure, is that an issue?
